### DISCUSSION TITLE
Can we prioritize certain POI in the search function?

### DISCUSSION THREAD

**Discussion by jfwerner** [osmandapp/OsmAnd#13223] [Ideas]
Sorry if this has been posted before, I couldn't find any posts similar to this.

Not sure about how or what most people search, but for me most of the time I'm just looking for a town or a city, and not very often for a street. Despite that, when I search "München" for example, the first result is a street with the name "München" in it, which is in a small town far away. This is even worse for small villages, roads with their names in it will display first.

**Can we not prioritize cities and towns more than roads?** I realize the custom search exists, but it's inconvenient and doesn't work for some reason in the above examples (maps aren't downlaoded, so the search isn't online maybe?)

Anyways, appreciate your work.

**Comment by vshcherb** [comment]
We push streets first that they are available for incremental search though indeed "München" as 1 word and complete search should select city first. Mostly people focus on searching address like Street, Hno or City, Street Hno

**Reply by jfwerner** [reply]
Sure, but it's an exact character match. Wouldn't it make sense to make that the first search result?

**Reply by Lee-Carre** [reply]
A couple thoughts (besides @jfwerner's fine suggestion for doing efficient matching first, before broadening).

* while I'm sure that there's some good reason not to: start at the top level; there are fewer nations than states, fewer states than cities, fewer cities cities than streets. In the context of big-O notation of performance, fewer nodes should mean that the total time is less, even with an inefficient algorithm. This approach will search everything eventually, but focuses on the easy results first.
* instead of always doing a global search (from within local data), be somewhat context dependent; search for streets (etc.) within current area (city, a radius from user's position, or whatever), but a general (top-down?) search of other places. The streets of another city don't matter much until the user is in that city. Local streets are likely of more interest.
* an option to sort the results by best-match, rather than the (ordinarily sensible) default of distance from user
* if the above aren't viable for any reason, then maybe some kind of branching search. My inspiration comes from chess engines attempting to compute the optimal next move.

Though, a more granular search, like the OP seeks, is appealing. However, other systems exist for that; [Overpass Turbo](//www.overpass-turbo.eu/) comes to mind.

**Reply by vshcherb** [reply]
Didn't understand the point, all first pois / streets should be exact match. All search results are sorted by how many words they exactly match from input query but they are not sorted for extra words (I'm might be wrong with but clause)
